### PR TITLE
optimize: refactoring client request processing logic.

### PR DESCRIPTION
**Pull request by objcoding**
### Ⅰ. Describe what this PR did
1. add netty processor,  in order to separate the processing business from netty.
2. add client processor：
ClientHeartbeatProcessor
ClientOnResponseProcessor
RmBranchCommitProcessor
RmBranchRollbackProcessor
RmUndoLogProcessor
3. remove interfaces:
ClientMessageSender
RegisterMsgListener
ClientMessageListener
4. add RemotingBootstrap interface.
### Ⅱ. Does this pull request fix one issue?
refactoring tasks: https://github.com/seata/seata/issues/2313

### III.Processor call flow.
server -> client
image

### IV. TODO 
1. add server processor.
2. refactoring rpc request method.

### CODE REVIEW

File: core/src/main/java/io/seata/core/rpc/netty/processor/client/MergeResultMessageProcessor.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    /**
+     * The Merge msg map from AbstractRpcRemoting.
+     */
+    private Map<Integer, MergeMessage> mergeMsgMap;
+
+    /**
+     * The Futures from AbstractRpcRemoting.
+     */
+    private ConcurrentHashMap<Integer, MessageFuture> futures;
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/processor/client/MergeResultMessageProcessor.java:49]: interface type

File: core/src/main/java/io/seata/core/rpc/netty/processor/client/RmHandleBranchRollbackProcessor.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    private void handleBranchRollback(RpcMessage request, String serverAddress, BranchRollbackRequest branchRollbackRequest) {
+        BranchRollbackResponse resultMessage;
+        resultMessage = (BranchRollbackResponse) handler.onRequest(branchRollbackRequest, null);
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("branch rollback result:" + resultMessage);
+        }
+        try {
+            this.remotingClient.sendResponse(request, serverAddress, resultMessage);
+        } catch (Throwable throwable) {
+            LOGGER.error("send response error: {}", throwable.getMessage(), throwable);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/netty/processor/client/RmHandleBranchRollbackProcessor.java:70]: Why is the exception handling different from RmHandleBranchCommitProcessor?

File: rm/src/main/java/io/seata/rm/RMClient.java [hunk 3]
```diff
@@ -34,7 +44,32 @@
...
+        processorMap.put((int) MessageType.TYPE_BRANCH_ROLLBACK, branchRollbackProcessor);
+        // rm handler undo log processor
+        Pair<NettyProcessor, Boolean> deleteUndoLogProcessor =
+            new Pair<>(new RmHandleUndoLogProcessor(handler), true);
+        processorMap.put((int) MessageType.TYPE_RM_DELETE_UNDOLOG, deleteUndoLogProcessor);
+        // handle TC response about process merge message
+        Pair<NettyProcessor, Boolean> mergeMsgProcessor =
+            new Pair<>(new MergeResultMessageProcessor(rmRpcClient.getMergeMsgMap(), rmRpcClient.getFutures()), false);
+        processorMap.put((int) MessageType.TYPE_SEATA_MERGE_RESULT, mergeMsgProcessor);
+        rmRpcClient.setRmProcessor(processorMap);
```
Comment (slievrly) [inline rm/src/main/java/io/seata/rm/RMClient.java:66]: remove

File: tm/src/main/java/io/seata/tm/TMClient.java [hunk 4]
```diff
@@ -32,6 +40,18 @@
      */
     public static void init(String applicationId, String transactionServiceGroup) {
         TmRpcClient tmRpcClient = TmRpcClient.getInstance(applicationId, transactionServiceGroup);
+
+        Map<Integer, Pair<NettyProcessor, Boolean>> processorMap = new HashMap<>();
+        // handle TC response about process merge message
+        Pair<NettyProcessor, Boolean> mergeMsgProcessor =
+            new Pair<>(new MergeResultMessageProcessor(tmRpcClient.getMergeMsgMap(), tmRpcClient.getFutures()), false);
+        processorMap.put((int) MessageType.TYPE_SEATA_MERGE_RESULT, mergeMsgProcessor);
+        tmRpcClient.setTmProcessor(processorMap);
```
Comment (slievrly) [inline tm/src/main/java/io/seata/tm/TMClient.java:49]: remove

File: rm/src/main/java/io/seata/rm/RMClient.java [hunk 5]
```diff
@@ -34,7 +44,32 @@
     public static void init(String applicationId, String transactionServiceGroup) {
         RmRpcClient rmRpcClient = RmRpcClient.getInstance(applicationId, transactionServiceGroup);
         rmRpcClient.setResourceManager(DefaultResourceManager.get());
-        rmRpcClient.setClientMessageListener(new RmMessageListener(DefaultRMHandler.get(), rmRpcClient));
+        AbstractRMHandler handler = DefaultRMHandler.get();
+
+        Map<Integer, Pair<NettyProcessor, Boolean>> processorMap = new HashMap<>();
+        // rm client handle branch commit processor
+        Pair<NettyProcessor, Boolean> branchCommitProcessor =
```
Comment (slievrly) [inline rm/src/main/java/io/seata/rm/RMClient.java:51]: Whether or not to use a user thread pool can be determined based on the request/response type?

File: core/src/main/java/io/seata/core/rpc/netty/TmRpcClient.java [hunk 6]
```diff
@@ -119,9 +124,19 @@ public void setApplicationId(String applicationId) {
     public void setTransactionServiceGroup(String transactionServiceGroup) {
         this.transactionServiceGroup = transactionServiceGroup;
     }
+
+    public void setTmProcessor(Map<Integer, Pair<NettyProcessor, Boolean>> processorMap) {
```
[the remark this answers is not in the document]
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/netty/TmRpcClient.java]: Agree with your suggestion, have been encapsulated.
```
io.seata.core.rpc.netty.TmRpcClient#registerProcessor
```

File: core/src/main/java/io/seata/core/rpc/netty/processor/client/MergeResultMessageProcessor.java (hunk 1 above)
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/netty/processor/client/MergeResultMessageProcessor.java]: Done.

File: rm/src/main/java/io/seata/rm/RMClient.java (hunk 3 above)
Comment (objcoding) [reply rm/src/main/java/io/seata/rm/RMClient.java]: Done.

File: tm/src/main/java/io/seata/tm/TMClient.java (hunk 4 above)
Comment (objcoding) [reply tm/src/main/java/io/seata/tm/TMClient.java]: Done.

File: rm/src/main/java/io/seata/rm/RMClient.java (hunk 5 above)
Comment (objcoding) [reply rm/src/main/java/io/seata/rm/RMClient.java]: Phase4 pr has already done so, I now migrate the phase4 related logic to the current pr.

File: core/src/main/java/io/seata/core/rpc/netty/processor/Pair.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ */
+package io.seata.core.rpc.netty.processor;
+
+/**
+ * Currently used to associate object1 and object2
+ *
+ * @author zhangchenghui.dev@gmail.com
+ * @since 1.2.0
+ */
+public class Pair<T1, T2> {
```
[the remark this answers is not in the document]
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/netty/processor/Pair.java]: Done.

File: core/src/main/java/io/seata/core/rpc/netty/processor/client/RmHandleBranchRollbackProcessor.java (hunk 2 above)
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/netty/processor/client/RmHandleBranchRollbackProcessor.java]: This is the old code. 
There are BranchCommitResponse and BranchRollbackResponse for customer response to the server request, but the server does anything to the response request, so I think the exception handling of RmBranchCommitProcessor is meaningless, so I will change it to the same as RmBranchRollbackProcessor .

Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/netty/processor/client/RmHandleBranchRollbackProcessor.java]: Done.

**Comment by slievrly** [comment]
checkstyle

**Comment by objcoding** [comment]
> checkstyle

ok.

**Comment by a364176773** [comment]
please check the test case:
StateMachineDBTests.testCompensationAndSubStateMachineAsyncWithLayout:563

**Comment by objcoding** [comment]
> testCompensationAndSubStateMachineAsyncWithLayout

Unit tests are OK.

File: core/src/main/java/io/seata/core/rpc/processor/client/RmBranchCommitProcessor.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        Object msg = rpcMessage.getBody();
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info("rm client handle branch commit process:" + msg);
+        }
+        handleBranchCommit(rpcMessage, remoteAddress, (BranchCommitRequest) msg);
+    }
+
+    private void handleBranchCommit(RpcMessage request, String serverAddress, BranchCommitRequest branchCommitRequest) {
+        BranchCommitResponse resultMessage;
+        resultMessage = (BranchCommitResponse) handler.onRequest(branchCommitRequest, null);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/processor/client/RmBranchCommitProcessor.java:63]: If need to handle exceptions on the outer layer here?

File: core/src/main/java/io/seata/core/rpc/processor/client/RmBranchRollbackProcessor.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        Object msg = rpcMessage.getBody();
+        if (LOGGER.isInfoEnabled()) {
+            LOGGER.info("rm handle branch rollback process:" + msg);
+        }
+        handleBranchRollback(rpcMessage, remoteAddress, (BranchRollbackRequest) msg);
+    }
+
+    private void handleBranchRollback(RpcMessage request, String serverAddress, BranchRollbackRequest branchRollbackRequest) {
+        BranchRollbackResponse resultMessage;
+        resultMessage = (BranchRollbackResponse) handler.onRequest(branchRollbackRequest, null);
```
Comment (slievrly) [inline core/src/main/java/io/seata/core/rpc/processor/client/RmBranchRollbackProcessor.java:63]: same as above

File: core/src/main/java/io/seata/core/rpc/processor/client/RmBranchCommitProcessor.java (hunk 8 above)
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/processor/client/RmBranchCommitProcessor.java:63]: This code does require active trycache:
```
handler.onRequest(branchCommitRequest, null)
```

Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/processor/client/RmBranchCommitProcessor.java:63]: Trycatch has been done in the following method：
```
io.seata.core.rpc.netty.AbstractNettyRemoting#processMessage
```

File: core/src/main/java/io/seata/core/rpc/processor/client/RmBranchRollbackProcessor.java (hunk 9 above)
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/processor/client/RmBranchRollbackProcessor.java:63]: Trycatch has been done in the following method：
```
io.seata.core.rpc.netty.AbstractNettyRemoting#processMessage
```

File: core/src/main/java/io/seata/core/rpc/processor/client/ClientOnResponseProcessor.java [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+                    if (LOGGER.isInfoEnabled()) {
+                        LOGGER.info("msg: {} is not found in futures.", msgId);
+                    }
+                } else {
+                    future.setResultMessage(results.getMsgs()[i]);
+                }
+            }
+        } else {
+            MessageFuture messageFuture = null;
+            if (!(rpcMessage.getBody() instanceof MergeResultMessage)) {
```
Comment (l81893521) [inline core/src/main/java/io/seata/core/rpc/processor/client/ClientOnResponseProcessor.java:109]: I think this 'if' can be removed.

File: core/src/main/java/io/seata/core/rpc/netty/RmRpcClient.java [hunk 11]
```diff
@@ -247,4 +256,27 @@ public String getMergedResourceKeys() {
...
+        ClientOnResponseProcessor onResponseProcessor =
+            new ClientOnResponseProcessor(mergeMsgMap, super.getFutures(), getTransactionMessageHandler());
+        registerProcessor(MessageType.TYPE_SEATA_MERGE_RESULT, onResponseProcessor, null);
+        registerProcessor(MessageType.TYPE_BRANCH_REGISTER_RESULT, onResponseProcessor, null);
+        registerProcessor(MessageType.TYPE_BRANCH_STATUS_REPORT_RESULT, onResponseProcessor, null);
+        registerProcessor(MessageType.TYPE_GLOBAL_LOCK_QUERY_RESULT, onResponseProcessor, null);
+        registerProcessor(MessageType.TYPE_REG_RM_RESULT, onResponseProcessor, null);
+        // 5.registry heartbeat message processor
+        ClientHeartbeatProcessor clientHeartbeatProcessor = new ClientHeartbeatProcessor();
+        super.registerProcessor(MessageType.TYPE_HEARTBEAT_MSG, clientHeartbeatProcessor, null);
```
Comment (l81893521) [inline core/src/main/java/io/seata/core/rpc/netty/RmRpcClient.java:280]: why some registerProcessor have 'super.' and some have not.

File: core/src/main/java/io/seata/core/rpc/processor/client/ClientOnResponseProcessor.java (hunk 10 above)
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/processor/client/ClientOnResponseProcessor.java]: Done.

File: core/src/main/java/io/seata/core/rpc/netty/RmRpcClient.java (hunk 11 above)
Comment (objcoding) [reply core/src/main/java/io/seata/core/rpc/netty/RmRpcClient.java:280]: Done.

**Pull request merged**